software/webapp/src/main/java/org/apache/brooklyn/entity/webapp/JavaWebAppMigrateEffector.java

```diff
@@ -0,0 +1,115 @@
+                return null;
+            }
+        });
+    }
+
+    private static void migrate(EntityInternal entity, String locationSpec) {
+        if (ServiceStateLogic.getExpectedState(entity) != Lifecycle.RUNNING) {
+            // Is it necessary to check if the whole application is healthy?
+            throw new RuntimeException("The entity needs to be healthy before the migration starts");
+        }
+
+        if (entity.getParent() != null && !entity.getParent().equals(entity.getApplication())) {
+            /*
+             * TODO: Allow nested entites to be migrated
+             * If the entity has a parent different to the application root the migration cannot be done right now,
+             * as it could lead into problems to deal with hierarchies like SameServerEntity -> Entity
+             */
+
+            throw new RuntimeException("Nested entities cannot be migrated right now");
+        }
+
+        // Retrieving the location from the catalog.
+        Location newLocation = Entities.getManagementContext(entity).getLocationRegistry().resolve(locationSpec);
+
+        // TODO: Find a better way to check if you're migrating an entity to the exactly same VM. This not always works.
+        for (Location oldLocation : entity.getLocations()) {
+            if (oldLocation.containsLocation(newLocation)) {
+                LOG.warn("You cannot migrate an entity to the same location, the migration process will stop right now");
+                return;
+            }
+        }
+
+        LOG.info("Migration process of " + entity.getId() + " started.");
+
+        // When we have the new location, we free the resources of the current instance
+        entity.invoke(Startable.STOP, MutableMap.<String, Object>of()).blockUntilEnded();
+
+        // Clearing old locations to remove the relationship with the previous instance
+        entity.clearLocations();
+        entity.addLocations(Lists.newArrayList(newLocation));
+
+        // Starting the new instance
+        entity.invoke(Startable.START, MutableMap.<String, Object>of()).blockUntilEnded();
+
+        // Refresh all the dependent entities
+
+        for (Entity applicationChild : entity.getApplication().getChildren()) {
+            // TODO: Find a better way to refresh the application configuration.
+            // TODO: Refresh nested entities or find a way to propagate the restart properly.
+
+            // Restart any entity but the migrated one.
+            if (applicationChild instanceof Startable) {
+                if (entity.equals(applicationChild)) {
+                    // The entity is sensors should rewired automatically on stop() + restart()
+                } else {
+                    // Restart the entity to fetch again all the dependencies (ie. attributeWhenReady ones)
+                    ((Startable) applicationChild).restart();
```

Comment:
We're not going to know which entities depend on each other. As I tried to describe in my first comment we must configure the other entities to behave sensibly when one is migrated.

The guiding principle to follow is that entities should act on information they receive (i.e. sensor events), not be unilaterally acted upon as is being done here.

Incidentally, support for more general relationships is making its way into Brooklyn. See @ahgittin's recent work in #987 for the foundations.